#include "gamewindow.h"
#include "replaywindow.h"
#include "ui_gamewindow.h"
#include <QVBoxLayout>
#include <QMessageBox>
#include <QRandomGenerator>
#include <QDebug>

Gamewindow::Gamewindow(int boardSize, int obstacleCount, const QString &firstPlayer, QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::Gamewindow)
    , boardSize(boardSize)
    , obstacleCount(obstacleCount)
    , firstPlayer(firstPlayer)
{
    ui->setupUi(this);

    // Î≥¥Îìú ÏÉùÏÑ±
    board = new OtheloBoard(this);
    board->setMinimumSize(400, 400);
    board->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Preferred);

    // Í∏∞Ï°¥ layout Ï†úÍ±∞
    if (ui->centralwidget->layout()) {
        delete ui->centralwidget->layout();
    }

    // Î©îÏù∏ Î†àÏù¥ÏïÑÏõÉ: ÏàòÌèâ (Ï¢å: Î≥¥Îìú / Ïö∞: ÏÉÅÌÉúÌëúÏãú + Î≤ÑÌäº)
    QHBoxLayout *mainLayout = new QHBoxLayout(ui->centralwidget);

    // ÏôºÏ™Ω: Î≥¥Îìú Î†àÏù¥ÏïÑÏõÉ
    QVBoxLayout *boardLayout = new QVBoxLayout();
    boardLayout->addWidget(board);
    mainLayout->addLayout(boardLayout, 3);  // ÎπÑÏú® Ï°∞Ï†ï

    // Ïò§Î•∏Ï™Ω: ÏÉÅÌÉú ÌëúÏãúÏ∞Ω + Î≤ÑÌäº
    QVBoxLayout *sideLayout = new QVBoxLayout();
    sideLayout->addWidget(ui->turnLabel);
    sideLayout->addWidget(ui->countLabel);
    sideLayout->addStretch();  // Ï§ëÍ∞Ñ Îπà Í≥µÍ∞Ñ
    sideLayout->addWidget(ui->undoButton);
    sideLayout->addWidget(ui->mainMenuButton);
    mainLayout->addLayout(sideLayout, 1);

    // Î≥¥Îìú ÌÅ¥Î¶≠ Ïãú ÎèôÏûë Ïó∞Í≤∞
    connect(board, &OtheloBoard::cellClicked, this, &Gamewindow::handleCellClicked);
    connect(ui->undoButton, &QPushButton::clicked, this, &Gamewindow::restorePreviousState);
    connect(ui->mainMenuButton, &QPushButton::clicked, this, &Gamewindow::onMainMenuClicked);

    // Ï¥àÍ∏∞ Î≥¥Îìú ÏÑ∏ÌåÖ
    initializeBoardUI();
    updateStatus();

    this->setMinimumSize(800, 700);  // Ï∞Ω ÏµúÏÜå ÌÅ¨Í∏∞ ÏÑ§Ï†ï
}

Gamewindow::~Gamewindow() {
    delete ui;
}

void Gamewindow::initializeBoardUI() {
    boardState.resize(boardSize, QVector<int>(boardSize, EMPTY));

    int mid = boardSize / 2;
    boardState[mid - 1][mid - 1] = WHITE;
    boardState[mid - 1][mid]     = BLACK;
    boardState[mid][mid - 1]     = BLACK;
    boardState[mid][mid]         = WHITE;

    placeObstacles();

    board->setBoardSize(boardSize);
    board->setBoard(boardState);

    updateValidMoves();
    board->setValidMoves(validMoves);  // üî∏ Ï§ëÏöî: Ï¥àÎ°ù ÌëúÏãú
}

void Gamewindow::handleCellClicked(int row, int col) {
    int currentPlayer = (firstPlayer == "Ìùë") ? BLACK : WHITE;
    if (!isValidMove(row, col, currentPlayer)) return;

    saveCurrentState();
    placeStone(row, col, currentPlayer);
    flipStones(row, col, currentPlayer);

    updateValidMoves();
    updateBoardUI();
    updateStatus();
    switchTurn();
}

void Gamewindow::updateBoardUI() {
    board->setBoard(boardState);
    board->setValidMoves(validMoves);  // üî∏ ÏóÜÏúºÎ©¥ Ï¥àÎ°ù ÏÇ¨ÎùºÏßê
}

bool Gamewindow::isValidMove(int row, int col, int currentPlayer) const {
    if (boardState[row][col] != EMPTY) return false;

    int opponent = (currentPlayer == BLACK) ? WHITE : BLACK;
    const int dx[] = {-1, -1, -1, 0, 1, 1, 1, 0};
    const int dy[] = {-1,  0,  1, 1, 1, 0, -1, -1};

    for (int d = 0; d < 8; ++d) {
        int x = row + dx[d];
        int y = col + dy[d];
        bool hasOpponentBetween = false;

        while (x >= 0 && x < boardSize && y >= 0 && y < boardSize) {
            if (boardState[x][y] == opponent) {
                hasOpponentBetween = true;
            } else if (boardState[x][y] == currentPlayer && hasOpponentBetween) {
                return true;
            } else {
                break;
            }
            x += dx[d];
            y += dy[d];
        }
    }
    return false;
}

void Gamewindow::placeStone(int row, int col, int currentPlayer) {
    boardState[row][col] = currentPlayer;
    updateStatus();

    ReplayState state;
    state.board = boardState;
    state.move = {row, col, currentPlayer};
    replayStates.append(state);
}

void Gamewindow::flipStones(int row, int col, int currentPlayer) {
    int opponent = (currentPlayer == BLACK) ? WHITE : BLACK;
    const int dx[] = {-1, -1, -1, 0, 1, 1, 1, 0};
    const int dy[] = {-1,  0,  1, 1, 1, 0, -1, -1};

    for (int d = 0; d < 8; ++d) {
        int x = row + dx[d];
        int y = col + dy[d];
        QVector<QPair<int, int>> toFlip;

        while (x >= 0 && x < boardSize && y >= 0 && y < boardSize) {
            if (boardState[x][y] == opponent) {
                toFlip.append({x, y});
            } else if (boardState[x][y] == currentPlayer) {
                for (auto &[fx, fy] : toFlip) {
                    boardState[fx][fy] = currentPlayer;
                }
                break;
            } else {
                break;
            }
            x += dx[d];
            y += dy[d];
        }
    }
}

void Gamewindow::switchTurn() {
    firstPlayer = (firstPlayer == "Ìùë") ? "Î∞±" : "Ìùë";
    updateValidMoves();

    if (validMoves.isEmpty()) {
        qDebug() << firstPlayer << "ÏùÄ Ï∞©ÏàòÌï† Ïàò ÏóÜÏñ¥ Ìå®Ïä§Îê©ÎãàÎã§.";
        firstPlayer = (firstPlayer == "Ìùë") ? "Î∞±" : "Ìùë";
        updateValidMoves();

        if (validMoves.isEmpty()) {
            checkGameOver();
            return;
        } else {
            QMessageBox::information(this, "Ìå®Ïä§", "Ï∞©Ïàò Í∞ÄÎä•Ìïú ÏúÑÏπòÍ∞Ä ÏóÜÏñ¥ Ìå®Ïä§Îê©ÎãàÎã§.");
        }
    }

    updateBoardUI();
    updateStatus();
    checkGameOver();
    qDebug() << "ÌÑ¥ Ï†ÑÌôò ‚Üí ÌòÑÏû¨ ÌÑ¥:" << firstPlayer;
}

void Gamewindow::updateValidMoves() {
    validMoves.clear();
    int currentPlayer = (firstPlayer == "Ìùë") ? BLACK : WHITE;

    for (int row = 0; row < boardSize; ++row) {
        for (int col = 0; col < boardSize; ++col) {
            if (isValidMove(row, col, currentPlayer)) {
                validMoves.append({row, col});
            }
        }
    }

    if (board) board->setValidMoves(validMoves);  // üî∏ Ï§ëÏöî
}

void Gamewindow::updateStatus() {
    if (!ui || !ui->turnLabel || !ui->countLabel) return;

    int blackCount = 0, whiteCount = 0;
    for (const auto &row : boardState) {
        for (int cell : row) {
            if (cell == BLACK) ++blackCount;
            else if (cell == WHITE) ++whiteCount;
        }
    }

    ui->turnLabel->setText(QString("ÌòÑÏû¨ ÌÑ¥: %1").arg(firstPlayer));

    // üî∂ ÌòÑÏû¨ ÌÑ¥ Î∞∞Í≤ΩÏÉâ Í∞ïÏ°∞
    if (firstPlayer == "Ìùë") {
        ui->turnLabel->setStyleSheet("background-color: black; color: white; font-weight: bold; font-size: 16px;");
    } else {
        ui->turnLabel->setStyleSheet("background-color: white; color: black; font-weight: bold; font-size: 16px;");
    }

    ui->countLabel->setText(QString("Ìùë: %1, Î∞±: %2").arg(blackCount).arg(whiteCount));
}

void Gamewindow::checkGameOver() {
    int blackCount = 0, whiteCount = 0;
    for (const auto &row : boardState) {
        for (int cell : row) {
            if (cell == BLACK) ++blackCount;
            else if (cell == WHITE) ++whiteCount;
        }
    }

    if (hasValidMove(BLACK) || hasValidMove(WHITE)) return;

    // Í≤åÏûÑ Í≤∞Í≥º ÏïåÎ¶º
    QString result;
    if (blackCount > whiteCount) result = "Ìùë ÏäπÎ¶¨!";
    else if (whiteCount > blackCount) result = "Î∞± ÏäπÎ¶¨!";
    else result = "Î¨¥ÏäπÎ∂ÄÏûÖÎãàÎã§!";
    QMessageBox::information(this, "Í≤åÏûÑ Ï¢ÖÎ£å", result);

    // Î¶¨ÌîåÎ†àÏù¥ Ï†ÄÏû• Ïó¨Î∂Ä ÌôïÏù∏
    bool saved = false;
    auto reply = QMessageBox::question(this, "Î¶¨ÌîåÎ†àÏù¥ Ï†ÄÏû•", "Î¶¨ÌîåÎ†àÏù¥Î•º Ï†ÄÏû•ÌïòÏãúÍ≤†ÏäµÎãàÍπå?");
    if (reply == QMessageBox::Yes) {
        saved = true;
    }
    gameEndedFlag = true;
    emit gameEnded(replayStates, boardSize, saved);  // ‚úÖ Ï†ÄÏû• Ïó¨Î∂Ä Ï†ÑÎã¨

    this->hide();
    parentWidget()->show();
}


bool Gamewindow::hasValidMove(int playerColor) const {
    for (int row = 0; row < boardSize; ++row) {
        for (int col = 0; col < boardSize; ++col) {
            if (isValidMove(row, col, playerColor)) {
                return true;
            }
        }
    }
    return false;
}

void Gamewindow::placeObstacles() {
    int placed = 0;
    int pairsToPlace = obstacleCount / 2;
    int mid = boardSize / 2;

    QSet<QPair<int, int>> forbidden = {
        {mid - 1, mid - 1}, {mid - 1, mid},
        {mid,     mid - 1}, {mid,     mid}
    };
    QSet<QPair<int, int>> used;

    int attempts = 0;
    const int maxAttempts = 1000;

    while (placed < pairsToPlace && attempts < maxAttempts) {
        int row = QRandomGenerator::global()->bounded(boardSize);
        int col = QRandomGenerator::global()->bounded(boardSize);
        int symRow = row;
        int symCol = boardSize - 1 - col;
        attempts++;

        QPair<int, int> p1 = {row, col};
        QPair<int, int> p2 = {symRow, symCol};

        if (forbidden.contains(p1) || forbidden.contains(p2) ||
            used.contains(p1) || used.contains(p2) ||
            boardState[row][col] != EMPTY || boardState[symRow][symCol] != EMPTY)
            continue;

        boardState[row][col] = OBSTACLE;
        boardState[symRow][symCol] = OBSTACLE;
        used.insert(p1);
        used.insert(p2);
        placed++;
    }

    if (placed < pairsToPlace) {
        qDebug() << "‚ö†Ô∏è Ïû•Ïï†Î¨º ÏùºÎ∂ÄÎßå Î∞∞ÏπòÎê®. ÏöîÏ≤≠:" << obstacleCount << " Ïã§Ï†ú:" << placed * 2;
    }
}

void Gamewindow::saveCurrentState() {
    GameState state;
    state.board = boardState;
    state.turn = firstPlayer;
    undoStack.push(state);
}

void Gamewindow::restorePreviousState() {
    if (undoStack.isEmpty()) return;

    GameState state = undoStack.pop();
    boardState = state.board;
    firstPlayer = state.turn;

    updateValidMoves();
    updateBoardUI();
    updateStatus();

    // ‚úÖ ÎêòÎèåÎ¶¨Í∏∞ Ïãú Î¶¨ÌîåÎ†àÏù¥ÎèÑ Ìïú ÌÑ¥ Ï†úÍ±∞
    if (!replayStates.isEmpty()) {
        replayStates.removeLast();
    }
}


void Gamewindow::onMainMenuClicked() {
    if (!gameEndedFlag) {
        emit requestSave(this);  // Í≤åÏûÑ Ï§ëÏùº ÎïåÎßå Ïù¥Ïñ¥ÌïòÍ∏∞ Ï†ÄÏû•
    }
    this->hide();
    parentWidget()->show();
}
bool Gamewindow::isGameEnded() const {
    return gameEndedFlag;
}
